samples/nvidia-resnet/pipeline/pipeline.py

```diff
@@ -0,0 +1,115 @@
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+import kfp.dsl as dsl
+import datetime
+import os
+from kubernetes import client as k8s_client
+
+
+PIPELINE_NAME = 'resnet_cifar10_pipeline'
+
+
+def preprocess_op(persistent_volume_path, input_dir, output_dir, step_name='preprocess'):
+    return dsl.ContainerOp(
+        name=step_name,
+        image='nvcr.io/nvidian/sae/ananths:kubeflow-preprocess',
+        command=['python'],
+        arguments=[
+            '/scripts/preprocess.py',
+            '--input_dir', '%s/%s' % (persistent_volume_path, input_dir),
+            '--output_dir', '%s/%s' % (persistent_volume_path, output_dir),
+        ],
+        file_outputs={}
+    )
+
+
+def train_op(persistent_volume_path, input_dir, output_dir, step_name='train'):
+    return dsl.ContainerOp(
+        name=step_name,
+        image='nvcr.io/nvidian/sae/ananths:kubeflow-train',
+        command=['python'],
+        arguments=[
+            '/scripts/train.py',
+            '--input_dir', '%s/%s' % (persistent_volume_path, input_dir),
+            '--output_dir', '%s/%s' % (persistent_volume_path, output_dir),
+        ],
+        file_outputs={}
+    )
+
+
+def serve_op(persistent_volume_path, input_dir, step_name='serve'):
+    return dsl.ContainerOp(
+        name=step_name,
+        image='nvcr.io/nvidian/sae/ananths:kubeflow-serve',
+        command=['python3'],
+        arguments=[
+            '/scripts/serve.py',
+            '--input_dir', '%s/%s' % (persistent_volume_path, input_dir),
+        ],
+        file_outputs={}
+    )
+
+
+@dsl.pipeline(
+    name=PIPELINE_NAME,
+    description='Demonstrate the ResNet50 predict.'
+)
+def resnet_pipeline(
+    raw_data_dir=dsl.PipelineParam(name='raw-data-dir', value='raw_data'),
+    processed_data_dir=dsl.PipelineParam(name='processed-data-dir', value='processed_data'),
+    model_dir=dsl.PipelineParam(name='saved-model-dir', value='saved_model')
+):
+
+    op_dict = {}
+
+    persistent_volume_name = 'nvidia-workspace'
+    persistent_volume_path = '/mnt/workspace/'
+
+    op_dict['preprocess'] = preprocess_op(
+        persistent_volume_path, raw_data_dir, processed_data_dir)
+
+    op_dict['train'] = train_op(
+        persistent_volume_path, processed_data_dir, model_dir)
+    op_dict['train'].after(op_dict['preprocess'])
```

Comment:
Having to manually specify dependencies using `.after` is usually a symptom of components that are not fully composable. Making components composable is a bit complicated now due to the lack of artifact passing support, but it's still pretty easy to do.

If some component produces some data that another component will use, this data dependency must be made explicit: The producer component must have an output a reference to which is then passed to the consumer component as input.

`preprocess_op` should output `processed_data_dir` which will be passed to `train_op`
`train_op` should output `model_dir` which will be passed to `serve_op`

Reply:
Data dependency is now made explicit (i.e. passing output between ops instead of using `.after`)